src/java/org/broadinstitute/dropseqrna/metagene/UMIMetaGeneCollection.java

```diff
@@ -0,0 +1,408 @@
+		Set<MetaGene> unambiguous = new HashSet<>();
+
+		GroupingIterator<SAMRecord> gIter = new GroupingIterator<>(informativeReads.iterator(), new ReadNameComparator());
+
+		// process data by read name.  One read name with 1 or more reads is a ReadGroupResult
+		for (List<SAMRecord> list : gIter) {
+			ReadGroupResult r = getMetaGeneType (list);
+			MetaGene mg = new MetaGene(r.getGenes());
+			if (r.getType().equals(ReadGroupResult.MetaGeneTypeEnum.UNAMBIGIOUS))
+				unambiguous.add(mg);
+			if (r.getType().equals(ReadGroupResult.MetaGeneTypeEnum.AMBIGUOUS))
+				ambiguous.add(mg);
+		}
+		// these are mutually exclusive.
+		this.unambiguousMetaGenes=unambiguous;
+		this.ambiguousMetaGenes=ambiguous;
+		// pretty well validated, no need to waste time on the call.
+		// validate();
+		
+		try {
+			gIter.close();
+		} catch (IOException e) {
+			// This should never, ever throw an error.
+			e.printStackTrace();
+		}		
+	}
+
+	/**
+	 * Given some reads for this UMI have already been mapped uniquely to genes and share a read name, decide the state of these mappings.
+	 * 
+	 * Basic algorithm pseudocode:
+	 * 1) Get the unique gene names for this set of reads [this can be less than the number of reads in the list if reads are unmapped]
+	 * 2) If there's a single read, classify as unmapped [map quality=0], low MQ [low map quality], high map quality [unique] or other [missing classification status]
+	 * 3) If there's more than 1 read, and that read maps to multiple genes:
+	 * 		If 1 of the genes is uniquely mapped elsewhere, then this read supports the uniquely mapped one.
+	 * 		If none of the genes is uniquely mapped elsewhere, then this read is a meta gene.
+	 * 		If these reads map to multiple genes that map uniquely elsewhere, then this read is ambiguous.
+	 * @param list A list of SAMRecords to look at within the context of a Cell/UMI.
+	 * @return
+	 */
+	public ReadGroupResult getMetaGeneType (final List<SAMRecord> list) {
+		String readName = list.getFirst().getReadName();
+		Set<String> readGenes = new HashSet<>();
+
+		for (SAMRecord r: list) {
+			if (!r.getReadName().equals(readName)) throw new IllegalStateException("Read names don't match for reads that should all share the same name");
+			String gene = r.getStringAttribute(this.geneTag);
+			if (gene!=null) readGenes.add(gene);
+		}
+
+		List<String> orderedReadGenes = new ArrayList<>(readGenes);
+
+		// if there's a single read, the read is unmapped or unique, or I'm dumb and missed something (OTHER)
+		if (list.size()<2) {
+			int readMQ = list.getFirst().getMappingQuality();
+			if (readMQ==0) return new ReadGroupResult(orderedReadGenes, MetaGeneTypeEnum.UNMAPPED, list);
+			if (readMQ<this.mapQualityUnique) return new ReadGroupResult(orderedReadGenes, MetaGeneTypeEnum.LOW_MQ, list);
+			if (readMQ>=this.mapQualityUnique) return new ReadGroupResult(orderedReadGenes, MetaGeneTypeEnum.UNIQUE, list);
+			// TODO: Looks like a bug, because this code is never reached.
```

Comment:
@jamesnemesh It looks like OTHER is never returned.  Does that make sense?

Reply:
It looks like OTHER isn't needed, because the 3 cases capture all the possible options.  It's possible I was thinking about differentiating between unique and non-unique mappings, but found I didn't need it.  

If you want to remove the "OTHER" condition that's fine.